File: src/monitoring.js
// Visitor analytics and error monitoring, both built to keep children's data out.
//
// Analytics: Vercel Web Analytics (cookie-free). Loads only on the live site, only after the visitor
// presses "Accept" on the cookie notice, and only counts public pages (never lessons or dashboards).
// Errors: Sentry, only when VITE_SENTRY_DSN is set. No names, emails, IDs, tokens or page queries are sent.

export const CONSENT_KEY = 'ai_tutor_cookie_consent';
export const PUBLIC_PATHS = ['/', '/login', '/try-classroom', '/privacy', '/terms', '/cookies', '/join'];

const isPublic = (url) => {
  try {
    return PUBLIC_PATHS.includes(new URL(url, window.location.origin).pathname);
  } catch {
    return false;
  }
};

const hasConsent = () => {
  try {
    return localStorage.getItem(CONSENT_KEY) === 'all';
  } catch {
    return false;
  }
};

// Drops query strings (e.g. /join?code=...) and anything that is not a public page.
export const analyticsBeforeSend = (event) => {
  if (!event?.url || !isPublic(event.url)) return null;
  const url = new URL(event.url, window.location.origin);
  return { ...event, url: `${url.origin}${url.pathname}` };
};

let analyticsLoaded = false;
export const enableAnalytics = () => {
  if (analyticsLoaded || !import.meta.env.PROD || !hasConsent()) return;
  analyticsLoaded = true;
  window.va = window.va || function va(...args) {
    (window.vaq = window.vaq || []).push(args);
  };
  window.va('beforeSend', analyticsBeforeSend);
  const script = document.createElement('script');
  script.src = '/_vercel/insights/script.js';
  script.defer = true;
  document.head.appendChild(script);
};

// --- Sentry --------------------------------------------------------------------

const SECRET_KEYS = /token|password|authorization|cookie|email|first_name|last_name|user/i;

export const scrubEvent = (event) => {
  const clean = { ...event };
  delete clean.user;
  if (clean.request) {
    const request = { ...clean.request };
    delete request.cookies;
    delete request.data;
    delete request.headers;
    delete request.query_string;
    if (request.url) request.url = request.url.split('?')[0];
    clean.request = request;
  }
  if (clean.breadcrumbs) {
    clean.breadcrumbs = clean.breadcrumbs.map((b) => {
      const crumb = { ...b };
      if (crumb.data?.url) crumb.data = { ...crumb.data, url: String(crumb.data.url).split('?')[0] };
      if (crumb.message && SECRET_KEYS.test(crumb.message)) crumb.message = '[removed]';
      return crumb;
    });
  }
  return clean;
};

export const initSentry = () => {
  const dsn = import.meta.env.VITE_SENTRY_DSN;
  if (!dsn) return;
  import('@sentry/react')
    .then((Sentry) => {
      Sentry.init({
        dsn,
        environment: import.meta.env.VITE_VERCEL_ENV || import.meta.env.MODE,
        sendDefaultPii: false,
        tracesSampleRate: 0, // errors only: no performance tracing, no session replay
        beforeSend: scrubEvent,
      });
    })
    .catch(() => {
      // Monitoring must never break the app.
    });
};

File: backend/monitoring.py
"""Error monitoring (Sentry) and the system status shown to the admin.

Sentry starts only when SENTRY_DSN is set. Events are scrubbed before they leave the server:
no request bodies, cookies, headers, query strings or user details, so children's data never
reaches Sentry.
"""
import os

SECRET_WORDS = ("token", "password", "authorization", "cookie", "email", "username", "first_name", "last_name")


def scrub_event(event, hint=None):
    event.pop("user", None)
    request = event.get("request")
    if isinstance(request, dict):
        for key in ("data", "cookies", "headers", "query_string", "env"):
            request.pop(key, None)
        if isinstance(request.get("url"), str):
            request["url"] = request["url"].split("?")[0]
    # Local variables in stack frames can hold passwords or tokens.
    for exc in (event.get("exception") or {}).get("values") or []:
        for frame in (exc.get("stacktrace") or {}).get("frames") or []:
            frame.pop("vars", None)
    for crumb in (event.get("breadcrumbs") or {}).get("values") or []:
        message = str(crumb.get("message") or "")
        if any(word in message.lower() for word in SECRET_WORDS):
            crumb["message"] = "[removed]"
        crumb.pop("data", None)
    return event


def init_sentry():
    dsn = os.getenv("SENTRY_DSN")
    if not dsn:
        return False
    try:
        import sentry_sdk
        sentry_sdk.init(
            dsn=dsn,
            environment=os.getenv("VERCEL_ENV") or os.getenv("FLASK_ENV") or "development",
            release=(os.getenv("VERCEL_GIT_COMMIT_SHA") or "")[:12] or None,
            send_default_pii=False,
            include_local_variables=False,
            traces_sample_rate=0,
            before_send=scrub_event,
        )
        return True
    except Exception as e:  # monitoring must never stop the app
        print(f"Sentry could not start: {e}")
        return False


def system_status():
    """What is switched on (no secret values, only yes/no)."""
    return {
        "error_monitoring": bool(os.getenv("SENTRY_DSN")),
        "error_monitoring_frontend": bool(os.getenv("VITE_SENTRY_DSN")),
        "weekly_email": bool(os.getenv("RESEND_API_KEY")),
        "weekly_email_schedule": bool(os.getenv("CRON_SECRET")),
        "azure_voice": bool(os.getenv("SPEECH_KEY") and os.getenv("SPEECH_REGION")),
        "environment": os.getenv("VERCEL_ENV") or "local",
        "database_name": os.getenv("MONGO_DB_NAME", "smart_tutor"),
        "version": (os.getenv("VERCEL_GIT_COMMIT_SHA") or "")[:7] or "local",
    }
